### ISSUE TITLE
TLS Server

### ISSUE THREAD

**Issue by leobackes** [randombit/botan#2237]
Hello All,

I'm using Botan::TLS_Server on top of standalone ASIO.
I'm using a self-signed certificate as CA and another one as server certificate (signed with that self-signed one), both made with botan command line using ECDSA algorithm; session manager is Session_Manager_Noop, rng is AutoSeeded_RNG  and the policy is Strict_Policy.

I'm using version 2.11, and everything works fine on Windows and ArchLinux.

Now my question/problem:
If I upgrade botan on my ArchLinux, going to version 2.12.4 or when I built it on an Ubuntu 18.04 box (botan 2.4.0), I get an "alert TLS handshake failure." on the client side.

At the moment I have no idea what is wrong, or what else to look for.
If I use openssl as a client, I get the following:

```
openssl s_client -enable_pha -tls1_2 -state -msg -showcerts -connect localhost:8989 -CAfile /tmp/cert/tempcerts.pem

CONNECTED(00000003)
SSL_connect:before SSL initialization
>>> ??? [length 0005]
    16 03 01 00 db
>>> TLS 1.2, Handshake [length 00db], ClientHello
    01 00 00 d7 03 03 04 78 82 59 cd 66 a7 fe 49 17
    9c a6 cb b8 22 80 c9 d9 d3 cd f2 21 64 3d 8e 8b
    cd cd aa 5a 75 1a 00 00 38 c0 2c c0 30 00 9f cc
    a9 cc a8 cc aa c0 2b c0 2f 00 9e c0 24 c0 28 00
    6b c0 23 c0 27 00 67 c0 0a c0 14 00 39 c0 09 c0
    13 00 33 00 9d 00 9c 00 3d 00 3c 00 35 00 2f 00
    ff 01 00 00 76 00 00 00 1a 00 18 00 00 15 61 74
    65 6c 69 65 72 64 61 73 61 75 64 65 2e 65 73 70
    2e 62 72 00 0b 00 04 03 00 01 02 00 0a 00 0c 00
    0a 00 1d 00 17 00 1e 00 19 00 18 00 23 00 00 00
    16 00 00 00 17 00 00 00 0d 00 30 00 2e 04 03 05
    03 06 03 08 07 08 08 08 09 08 0a 08 0b 08 04 08
    05 08 06 04 01 05 01 06 01 03 03 02 03 03 01 02
    01 03 02 02 02 04 02 05 02 06 02
SSL_connect:SSLv3/TLS write client hello
<<< ??? [length 0005]
    15 03 03 00 02
<<< TLS 1.2, Alert [length 0002], fatal handshake_failure
    02 28
SSL3 alert read:fatal:handshake failure
SSL_connect:error in error
140040023487744:error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1543:SSL alert number 40
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 224 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1578354362
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```
Am I missing some initialization/configuring for other versions of botan?

**Comment by randombit** [comment]
I'm not sure about 2.4.0 both because that's quite old and because IIRC Debian/Ubuntu have various odd patches which may affect things. But if something worked for 2.11 it should certainly work with 2.12 as well, and the Arch build is basically stock without patches. I can't think of any recent changes that would affect usage as you describe.

When the alert is sent to the client, there should also be an exception on the server side that has a `.what()` value with some hopefully informative message, this may provide a hint.

It might be worth trying the default policy (`TLS::Policy`) instead of the strict policy though I don't think that is the issue here.

**Comment by leobackes** [comment]
Added some error handling on server, and indeed it had an exception on TLS_Server::received_data: `Can't agree on a ciphersuite with client`.

What seems weird to me is that the client I made with botan also uses the Strict policy.

I'll try with default policy and see what happens.
Thank you.

**Comment by randombit** [comment]
These issues can be hard to debug sometimes, but the fact that Botan is being used on both sides of the connection makes it quite strange. I can't replicate by generating an ECDSA cert (self-signed) then connecting to a `botan tls_server --policy=strict` instance using OpenSSL - it negotiates ECDHE_ECDSA_WITH_CHACHA20_POLY1305 as expected.

Unfortunately I'm going to be away the rest of this week and won't be able to look at it until next Sunday or so. In the meantime one thing to try would be taking 2.12.1 (or 2.13.0) and adding printfs to the function `choose_ciphersuite` in `tls_server.cpp` to track the flow. Especially print at the start `suite.to_string()` and then on each `continue` in the loop (these are where we give up on a particular ciphersuite)

Future enhancement - the "Can't agree" error message should be providing more information to help debugging (what the client offered, why we couldn't agree, etc)

**Comment by leobackes** [comment]
No worry, I'll be able to look this in more detail in the next month;
Then I'll built a debug version and get my GDB working on this :-)

I just opened this issue to be sure I wasn't skipping something obvious or some initialization.

By now, everything works on 2.11, so I'll keep this going as is for a while.

I'll come back as soon as I get something more.

**Comment by leobackes** [comment]
Hello again!

Back to this issue, after debugging the current master branch and the working version I have, I came to this:
At the client side, libbotan is trying to load the public key from the server.
It then reaches `std::unique_ptr<Public_Key> load_public_key(const AlgorithmIdentifier& alg_id, const std::vector<uint8_t>& key_bits)`, with the alg_id pointing to an OID "1.2.840.10.045.2.1".

On the working version, the code translates it with `OIDS::lookup(alg_id.get_oid())`, which returns "ECDSA" (the actual key type on the server-side).
On the current (not working) version, it calls `alg_id.get_oid().to_formatted_string();`, which doesn't make the translation, and returns "1.2.840.10.045.2.1", hence not matching with the "ECDSA" (nor any other algorithms), so it throws on pk_algs.cpp:159.

Looking at oid_maps.cpp, I realised  there is no "1.2.840.10.045.2.1", but "1.2.840.10045.2.1", so, I've came to the conclusion that somehow, somewhere inside OID::to_formatted_string() an extra dot is being inserted, as the vector m_id inside AlgorithmIdentifier is correctly `{1, 2, 840, 10045, 2, 1}`.
